hilti/toolchain/src/ast/ast-context.cc

```diff
@@ -45,8 +45,36 @@ ASTContext::ASTContext(Context* context) : _context(context) {
 }
 
 ASTContext::~ASTContext() {
-    for ( auto r : _nodes )
-        r->~Node();
+    try {
+        clear();
+    } catch ( const std::exception& e ) {
+        logger().internalError(util::fmt("unexpected exception in ~ASTContext: %s", e.what()));
+    }
+}
+
+void ASTContext::clear() {
+    if ( ! _root )
+        return;
+
```

Comment:
The body does not depend on `_root` being set, and the early exit makes the code more fragile. WDYT about dropping this?

```suggestion
```